tensorflow/compiler/jit/cluster_scoping_pass.cc

```diff
@@ -0,0 +1,148 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/jit/cluster_scoping_pass.h"
+
+#include "absl/algorithm/container.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+
+#include "tensorflow/compiler/jit/defs.h"
+#include "tensorflow/compiler/jit/xla_cluster_util.h"
+#include "tensorflow/core/framework/node_def_util.h"
+#include "tensorflow/core/graph/algorithm.h"
+
+namespace tensorflow {
+namespace {
+
+class ClusterScopingPassImpl {
+ public:
+  ClusterScopingPassImpl(Graph* graph,
+                         OptimizerOptions::GlobalJitLevel global_jit_level)
+      : graph_(graph),
+        global_jit_level_(global_jit_level),
+        unique_scope_id_(0) {}
+
+  Status Run();
+
+ private:
+  Status ScopingForPipelineStages();
+
+  size_t GetUniqueScopeId() { return unique_scope_id_++; }
+
+  void AddScopeToAllTransitivePredecessors(Node* start);
+
+  void AddScopeToAllTransitiveSuccessors(Node* start);
+
+ private:
+  Graph* graph_;
+  OptimizerOptions::GlobalJitLevel global_jit_level_;
+  size_t unique_scope_id_;
+};
+
+absl::optional<string> GetXlaInternalScope(Node* node) {
+  string scope;
+  if (GetNodeAttr(node->attrs(), kXlaInternalScopeAttr, &scope).ok()) {
+    return scope;
+  }
+
+  return absl::nullopt;
+}
+
+void SetXlaInternalScope(Node* node, StringPiece scope) {
+  node->AddAttr(kXlaInternalScopeAttr, scope);
+}
+
+// NB! We append a new scope as suffix to the _XlaInternalScope attribute
+// instead of overriding the old value.  In this way, we respect the original
+// scopes.  In other words, appending X to Y creates the conjunction of the
+// scopes X and Y (i.e, X & Y in effect).
+void AddOrAppendXlaInternalScope(Node* node, absl::string_view suffix) {
+  string updated_scope;
+  absl::optional<string> cur_scope = GetXlaInternalScope(node);
+  if (cur_scope == absl::nullopt) {
+    updated_scope = std::string(suffix);
+  } else {
+    updated_scope = absl::StrCat(cur_scope.value(), "&", suffix);
```

Comment:
Pedantically this is bit odd.  If we have `Unstage` -> `X` -> `Stage`; `Y` -> `Stage` then we'd be fine putting `X` and `Y` in the same cluster.  However, we won't do that in this scheme.

I think that's fine though (the situation above should be rare), but it would be nice to add a comment, if my understanding above is correct.

Reply:
See the above comment for the response.